openwisp_controller/api/serializers.py

```diff
@@ -0,0 +1,137 @@
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgOwned
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class TemplateSerializer(ValidatedModelSerializer):
+    config = serializers.JSONField()
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField()
+    organization = serializers.PrimaryKeyRelatedField(
+        allow_null=False, queryset=Organization.objects.all(), required=True
+    )
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(FilterSerializerByOrgOwned, ValidatedModelSerializer):
+    config = serializers.JSONField()
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = '__all__'
+
+
+class DeviceConfigSerializer(serializers.ModelSerializer):
+    config = serializers.JSONField()
+    context = serializers.JSONField()
+
+    class Meta:
+        model = Config
+        fields = ['backend', 'status', 'templates', 'context', 'config']
+
+
+class DeviceListSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer(write_only=True)
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = [
+            'id',
+            'name',
+            'organization',
+            'mac_address',
+            'key',
+            'last_ip',
+            'management_ip',
+            'model',
+            'os',
+            'system',
+            'notes',
+            'config',
+        ]
+
+    def create(self, validated_data):
+        config_data = validated_data.pop('config')
+        device = Device.objects.create(**validated_data)
+        config_templates = config_data.pop('templates')
+        config = Config.objects.create(device=device, **config_data)
+        for template in config_templates:
+            config.templates.add(template.pk)
+        return device
+
+
+class DeviceDetailSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer()
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = DeviceListSerializer.Meta.fields
+
+    def update(self, instance, validated_data):
+        config_data = None
+
+        if self.data['config'] is None and validated_data.get('config'):
+            config_data_ = dict(validated_data.get('config'))
```

Comment:
Can we change the variable name here to make it more descriptive to make it clear how `config_data_` is different from `config_data`?

Reply:
I have renamed it to `new_config_data`, this variable I am using when it is required to create a new config instance with a PUT/PATCH requests.